File: Game2D/Engine/Hero.cs
using System.Windows.Controls;
using System.Windows.Media.Imaging;
using System.Windows.Input;
using System.Collections.Generic;
using System.Windows;
using System.Windows.Media;

namespace Game2D.Engine
{
    public class Hero : GameObject
    {
        public int MaxHealth { get; set; } = 3000;
        public int Health { get; set; }
        public bool IsAlive => Health > 0;

        private int rifleAmmo = 30;
        private int maxRifleAmmo = 30;
        public int RifleAmmo => rifleAmmo;
        public int MaxRifleAmmo => maxRifleAmmo;

        public enum WeaponType { Rifle, Shotgun, Lasergun }
        private WeaponType currentWeapon = WeaponType.Rifle;
        private int shotgunAmmo = 5, maxShotgunAmmo = 5;
        private int lasergunAmmo = 10, maxLasergunAmmo = 10;
        private int shootCooldown = 0;
        private int rifleCooldown = 10, shotgunCooldown = 40, lasergunCooldown = 20;

        public Hero(double x, double y)
        {
            X = x;
            Y = y;
            Health = MaxHealth;
            Sprite = new Image
            {
                Source = new BitmapImage(new System.Uri("pack://application:,,,/Project/images/hero.png")),
                Width = 48,
                Height = 48
            };
        }

        // Проверка возможности движения по смещению dx, dy
        private bool CanMove(double dx, double dy)
        {
            var mainWindow = System.Windows.Application.Current.MainWindow as Game2D.MainWindow;
            if (mainWindow == null || mainWindow.GameCanvas == null) return false;
            double newX = X + dx;
            double newY = Y + dy;
            double maxX = mainWindow.GameCanvas.ActualWidth - Sprite.Width;
            double maxY = mainWindow.GameCanvas.ActualHeight - Sprite.Height;
            if (newX < 0 || newY < 0 || newX > maxX || newY > maxY)
                return false;
            // Убираем любые проверки на стены и препятствия
            return true;
        }

        public void MoveByKeys()
        {
            double speed = 2; // Было 4, теперь в 2 раза меньше
            double dx = 0, dy = 0;
    
            if (Keyboard.IsKeyDown(Key.W)) dy -= 1;
            if (Keyboard.IsKeyDown(Key.S)) dy += 1;
            if (Keyboard.IsKeyDown(Key.A)) dx -= 1;
            if (Keyboard.IsKeyDown(Key.D)) dx += 1;

            double length = Math.Sqrt(dx * dx + dy * dy);
            if (length > 0)
            {
                dx = dx / length * speed;
                dy = dy / length * speed;
            }

            if (CanMove(dx, dy))
            {
                X += dx;
                Y += dy;
            }
            else
            {
                if (dx != 0 && CanMove(dx, 0)) X += dx;
                if (dy != 0 && CanMove(0, dy)) Y += dy;
            }
        }

        public void ShootTo(double targetX, double targetY, GameWorld world)
        {
            double dx = targetX - (X + Sprite.Width / 2);
            double dy = targetY - (Y + Sprite.Height / 2);
            double angle = System.Math.Atan2(dy, dx);
            var canvas = world._canvas;
            switch (currentWeapon)
            {
                case WeaponType.Rifle:
                    if (rifleAmmo > 0 && shootCooldown == 0)
                    {
                        var bullet = new Bullet(canvas, X + Sprite.Width / 2, Y + Sprite.Height / 2, angle);
                        world.AddObject(bullet);
                        rifleAmmo--;
                        shootCooldown = rifleCooldown;
                    }
                    break;
                case WeaponType.Shotgun:
                    if (shotgunAmmo > 0 && shootCooldown == 0)
                    {
                        for (int i = -1; i <= 1; i++)
                        {
                            var bullet = new Bullet(canvas, X + Sprite.Width / 2, Y + Sprite.Height / 2, angle + i * 0.2);
                            world.AddObject(bullet);
                        }
                        shotgunAmmo--;
                        shootCooldown = shotgunCooldown;
                    }
                    break;
                case WeaponType.Lasergun:
                    if (lasergunAmmo > 0 && shootCooldown == 0)
                    {
                        var laser = new Laser(canvas, X + Sprite.Width / 2, Y + Sprite.Height / 2, angle);
                        world.AddObject(laser);
                        lasergunAmmo--;
                        shootCooldown = lasergunCooldown;
                    }
                    break;
            }
        }

        public void Reload()
        {
            switch (currentWeapon)
            {
                case WeaponType.Rifle:
                    rifleAmmo = maxRifleAmmo;
                    break;
                case WeaponType.Shotgun:
                    shotgunAmmo = maxShotgunAmmo;
                    break;
                case WeaponType.Lasergun:
                    lasergunAmmo = maxLasergunAmmo;
                    break;
            }
        }

        public void PickupWeapon(WeaponType type)
        {
            currentWeapon = type;
            switch (type)
            {
                case WeaponType.Rifle:
                    rifleAmmo = maxRifleAmmo;
                    break;
                case WeaponType.Shotgun:
                    shotgunAmmo = maxShotgunAmmo;
                    break;
                case WeaponType.Lasergun:
                    lasergunAmmo = maxLasergunAmmo;
                    break;
            }
        }

        public void SwitchWeapon(int weaponIndex)
        {
            switch (weaponIndex)
            {
                case 1:
                    currentWeapon = WeaponType.Rifle;
                    Sprite.Source = new BitmapImage(new System.Uri("pack://application:,,,/Project/images/hero.png"));
                    break;
                case 2:
                    currentWeapon = WeaponType.Shotgun;
                    Sprite.Source = new BitmapImage(new System.Uri("pack://application:,,,/Project/images/hero_shotgun.png"));
                    break;
                case 3:
                    currentWeapon = WeaponType.Lasergun;
                    Sprite.Source = new BitmapImage(new System.Uri("pack://application:,,,/Project/images/hero_lasergun.png"));
                    break;
            }
        }

        private void TurnTowardsMouse()
        {
            var mainWindow = System.Windows.Application.Current.MainWindow as Game2D.MainWindow;
            if (mainWindow == null || mainWindow.GameCanvas == null) return;
            var pos = System.Windows.Input.Mouse.GetPosition(mainWindow.GameCanvas);
            double centerX = X + (Sprite?.Width ?? 0) / 2;
            double centerY = Y + (Sprite?.Height ?? 0) / 2;
            double dx = pos.X - centerX;
            double dy = pos.Y - centerY;
            double angle = Math.Atan2(dy, dx) * 180 / Math.PI;
            if (Sprite != null)
                Sprite.RenderTransform = new System.Windows.Media.RotateTransform(angle, Sprite.Width/2, Sprite.Height/2);
        }

        public override void Update()
        {
            MoveByKeys();
            TurnTowardsMouse();
            if (shootCooldown > 0) shootCooldown--;
        }

        public WeaponType CurrentWeapon => currentWeapon;
        public int ShotgunAmmo => shotgunAmmo;
        public int MaxShotgunAmmo => maxShotgunAmmo;
        public int LasergunAmmo => lasergunAmmo;
        public int MaxLasergunAmmo => maxLasergunAmmo;

        public void ResetAmmo()
        {
            rifleAmmo = maxRifleAmmo;
            shotgunAmmo = maxShotgunAmmo;
            lasergunAmmo = maxLasergunAmmo;
        }
    }
} 

File: Game2D/Project/Hero.java
import greenfoot.*;

public class Hero extends Actor {
    private GreenfootImage rifle=new GreenfootImage("hero.png");
    private GreenfootImage shotgun=new GreenfootImage("hero_shotgun.png");
    private GreenfootImage lasergun=new GreenfootImage("hero_lasergun.png");
    
    private int direction;
    private boolean alive;
    
    private int MAX_HEALTH = 3000;
    public int health=MAX_HEALTH;

    private boolean isBuffed = false;
    private int buffTimer = 0;
    private final int buffDuration = 300;

    private int reloadTimer = 0;

    private int shootTimer = 0;

    private WeaponType currentWeapon = WeaponType.RIFLE;
    int weaponLevel=1;

    // Оружие и боезапас
    private int laserAmmo = 10, maxLaserAmmo = 10;
    private int shotgunAmmo = 5, maxShotgunAmmo = 5;
    private int rifleAmmo = 30, maxRifleAmmo = 30;

    private final int laserCooldown = 10;
    private final int shotgunCooldown = 120;
    private final int rifleCooldown = 80;

    public Hero(int startX, int startY, int startDirection) {
        setImageWithoutWhite("hero.png");

        this.direction = startDirection;
        this.alive = true;
        this.health = MAX_HEALTH;

        setRotation(direction * 90);
    }

    private void setImageWithoutWhite(String filename) {
        GreenfootImage image = new GreenfootImage(filename);
        for (int x = 0; x < image.getWidth(); x++) {
            for (int y = 0; y < image.getHeight(); y++) {
                if (image.getColorAt(x, y).equals(Color.WHITE)) {
                    image.setColorAt(x, y, new Color(0, 0, 0, 0));
                }
            }
        }
        setImage(image);
    }

    public void move(int dx, int dy) {
        if (!alive) return;

        int newX = getX() + dx;
        int newY = getY() + dy;

        if (canMove(dx, dy)) {
            setLocation(newX, newY);
        }
    }

    private boolean canMove(int dx, int dy) {
        return getOneObjectAtOffset(dx, dy+10, Wall_gorizont.class) == null && getOneObjectAtOffset(dx+10, dy, Wall_gorizont.class) == null &&
                getOneObjectAtOffset(dx, dy-10, Wall_gorizont.class) == null && getOneObjectAtOffset(dx-10, dy, Wall_gorizont.class) == null &&
               getOneObjectAtOffset(dx+10, dy, Wall_vertical.class) == null && getOneObjectAtOffset(dx, dy+10, Wall_vertical.class) == null &&
               getOneObjectAtOffset(dx-10, dy, Wall_vertical.class) == null && getOneObjectAtOffset(dx, dy-10, Wall_vertical.class) == null &&
               getOneObjectAtOffset(dx+10, dy, Cube.class) == null && getOneObjectAtOffset(dx-10, dy, Cube.class) == null &&
               getOneObjectAtOffset(dx, dy+10, Cube.class) == null && getOneObjectAtOffset(dx, dy-10, Cube.class) == null;
    }

    public void act() {
        if (!alive) return;

        handleBuff();
        turnTowardsMouse();
        
        switchWeapon();
        shoot();
        checkWeaponPickup();
        reload();
    }

    private void handleBuff() {
        if (isBuffed) {
            buffTimer++;
            if (buffTimer >= buffDuration) {
                isBuffed = false;
                buffTimer = 0;
            }
        }
    }

    private void switchWeapon() {
        if (Greenfoot.isKeyDown("3") && weaponLevel>=3){
            currentWeapon = WeaponType.LASERGUN;
            setImage(lasergun);
        }
        if (Greenfoot.isKeyDown("2") && weaponLevel>=2) {
            currentWeapon = WeaponType.SHOTGUN;
            setImage(shotgun);
        }
        if (Greenfoot.isKeyDown("1") && weaponLevel>=1) {
            currentWeapon = WeaponType.RIFLE;
            setImage(rifle);
        }
    }
    
    private void checkWeaponPickup() {
        WeaponPickup pickup = (WeaponPickup) getOneIntersectingObject(WeaponPickup.class);
        if (pickup != null) {
            currentWeapon = pickup.getWeaponType();
            switch (currentWeapon) {
            case LASERGUN:
                setImage(lasergun);
                break;
            case SHOTGUN:
                setImage(shotgun);
                break;
            case RIFLE:
                setImage(rifle);
                break;
        }
            weaponLevel++;
            getWorld().removeObject(pickup);
        }
    }

    public void shoot() {
        shootTimer++;
        if (!Greenfoot.mousePressed(null)) return;

        MouseInfo mouse = Greenfoot.getMouseInfo();
        if (mouse == null) return;

        int dx = mouse.getX() - getX();
        int dy = mouse.getY() - getY();
        double angle = Math.atan2(dy, dx);

        switch (currentWeapon) {
            case LASERGUN:
                if (shootTimer >= laserCooldown && laserAmmo > 0) {
                    fireLaser(angle);
                    laserAmmo--;
                    shootTimer = 0;
                }
                break;
            case SHOTGUN:
                if (shootTimer >= shotgunCooldown && shotgunAmmo > 0) {
                    for (int i = -1; i <= 1; i++) {
                        fireBullet(angle + i * 0.2);
                    }
                    shotgunAmmo--;
                    shootTimer = 0;
                }
                break;
            case RIFLE:
                if (shootTimer >= rifleCooldown && rifleAmmo > 0) {
                    fireBullet(angle);
                    rifleAmmo--;
                    shootTimer = 0;
                }
                break;
        }

        updateAmmoLabel();
    }

    private void fireBullet(double angle) {
        Bullet bullet = new Bullet(getX()+10, getY()+10, angle);
        getWorld().addObject(bullet, getX()+10, getY()+10);
    }
    
    private void fireLaser(double angle) {
        Laser laser = new Laser(angle);
        getWorld().addObject(laser, getX()+5, getY()+14);
    }

    public void reload() {
        reloadTimer++;
        if (reloadTimer >= (isBuffed ? 60 : 120)) {
            reloadTimer = 0;
            switch (currentWeapon) {
                case LASERGUN:
                    if (laserAmmo < maxLaserAmmo) laserAmmo++;
                    break;
                case SHOTGUN:
                    if (shotgunAmmo < maxShotgunAmmo) shotgunAmmo++;
                    break;
                case RIFLE:
                    if (rifleAmmo < maxRifleAmmo) rifleAmmo++;
                    break;
            }
            updateAmmoLabel();
        }
    }

    private void updateAmmoLabel() {
        MyWorld world = (MyWorld) getWorld();
        if (world != null) {
            int ammo = switch (currentWeapon) {
                case LASERGUN -> laserAmmo;
                case SHOTGUN -> shotgunAmmo;
                case RIFLE -> rifleAmmo;
            };
            world.updateAmmoDisplay(ammo);
        }
    }

    private void updateHealthLabel() {
        MyWorld world = (MyWorld) getWorld();
        if (world != null) {
            world.getHealthBar().updateHealth(health);
        }
    }

    public void takeDamage(int damage) {
        if (!alive) return;

        health -= damage;
        if (health < 0) health = 0;
        updateHealthLabel();

        if (health <= 0) die();
    }

    private void die() {
        alive = false;
        getWorld().addObject(new Died() , getWorld().getWidth()/2,getWorld().getHeight()/2);
        getWorld().removeObject(this);
        Greenfoot.stop();
    }
    
    public boolean isAlive(){
        return alive;
    }
    
    public int getMaxHealth(){
        return MAX_HEALTH;
    }

    public void applyBuff() {
        if (!isBuffed) {
            isBuffed = true;
            buffTimer = 0;
        }
    }

    private void turnTowardsMouse() {
        MouseInfo mouse = Greenfoot.getMouseInfo();
        if (mouse != null) {
            int dx = mouse.getX() - getX();
            int dy = mouse.getY() - getY();
            double angle = Math.atan2(dy, dx);
            setRotation((int) Math.toDegrees(angle));
        }
    }

    public boolean checkPortalCollision() {
        return getOneIntersectingObject(Portal.class) != null;
    }

    public boolean checkPortal2Collision() {
        return getOneIntersectingObject(Portal2.class) != null;
    }
    
    public boolean checkPortal3Collision() {
        return getOneIntersectingObject(Portal3.class) != null;
    }
}
